src/analyze/analyze-security.c

```diff
@@ -2100,6 +2113,138 @@ static int analyze_security_one(sd_bus *bus, const char *name, Table *overview_t
         return 0;
 }
 
+/* Refactoring SecurityInfo so that it can make use of existing struct variables instead of reading from dbus */
+static int get_security_info(Unit *u, ExecContext *c, CGroupContext *g, SecurityInfo **ret_info) {
+        _cleanup_(security_info_free_freep) SecurityInfo *info = new0(SecurityInfo, 1);
+        if (!info)
+                return log_oom();
+
+        assert(ret_info);
+
+        if (u) {
+                info->id = u->id;
```

Comment:
memory management still fucked. the id string will now be destroyed both by the destructor of "info" and the destructor of "u". This will result in memory corruption.

You need to copy everything, so that the into structure is fully self contained and owns all data it references.

Make sure to run your stuff through valgrind, it will tell you about many of these issues

Reply:
@bluca you suggested to use set_copy for syscall_archs but set_copy requires an identifier 'h' for it to work. So this command returns an error: info->system_call_architectures = set_copy(c->syscall_archs);

Reply:
Uh, yes indeed, good catch! There's no usage of that function anywhere so it was never caught. Please add an individual commit fixing it, and it would be very nice if it also included a testcase in src/test/test-set.c

Reply:
Fixed that but I still get segmentation faults for info->user and info->supplementary_groups. valgrind reveals that 'Access not within mapped region at address 0x0' and I don't know why this is happening. This is within get_security_info() in analyze-security.c and it references the strdup.c file so it seems like it has something to do with the duplications below:

        info->user = strdup(c->user);
                if (!info->user)
                        return log_oom();
        info->supplementary_groups = strv_copy(c->supplementary_groups);
                if (!info->supplementary_groups)
                        return log_oom();

Reply:
strdup comes from libc, and it's not happy if you pass NULL to it - in other words, check that it's non-null before strdup'ing it

Reply:
if that still doesn't work, push the current state and I'll have another look tomorrow morning